File: plugin/preset/elements/paragraph.lua
local Paragraph = {}
local Alignment = {
	LEFT = 0,
	CENTER = 1,
	RIGHT = 2,
}

function Paragraph:new(...)
	local o = {
		alignment = 0,
		position = nil,
		lines = { ... },
	}
	setmetatable(o, self)
	self.__index = self
	return o
end

function Paragraph:from(lines) return self:new(table.unpack(lines)) end

function Paragraph:align(align)
	self.alignment = align
	return self
end

function Paragraph:area(rect)
	self.position = rect
	return self
end

function Paragraph:to_string()
	local s = ""
	for _, line in ipairs(self.lines) do
		s = s .. line:to_string():gsub("\r", "\\\r") .. "\r"
	end
	return s.sub(s, 1, -2)
end

function Paragraph.render(...)
	local s = "R"
	for _, paragraph in ipairs { ... } do
		s = s
			.. paragraph.position.x
			.. ","
			.. paragraph.position.y
			.. ","
			.. paragraph.position.width
			.. ","
			.. paragraph.position.height
			.. ","
			.. paragraph.alignment
			.. ";"
			.. paragraph:to_string():gsub("\0", "\\\0")
			.. "\0"
	end
	return s.sub(s, 1, -2)
end

setmetatable(Paragraph, {
	__call = function(self, ...) return self:new(...) end,
	__tostring = function(self) return self:to_string() end,
})

ui = ui or {}
ui.Paragraph = Paragraph
ui.Alignment = Alignment

File: plugin/src/layout/paragraph.rs
use mlua::{AnyUserData, FromLua, Lua, Table, UserData, Value};
use ratatui::widgets::Widget;

use super::{Rect, Style};
use crate::{layout::Line, GLOBALS, LUA};

#[derive(Clone, Debug)]
pub struct Paragraph {
	area: ratatui::layout::Rect,

	text:      ratatui::text::Text<'static>,
	style:     Option<ratatui::style::Style>,
	alignment: ratatui::prelude::Alignment,
}

impl Paragraph {
	pub(crate) fn install() -> mlua::Result<()> {
		let ui: Table = GLOBALS.get("ui")?;
		ui.set(
			"Paragraph",
			LUA.create_function(|_, (area, lines): (Rect, Vec<Line>)| {
				Ok(Self {
					area: area.0,

					text:      lines.into_iter().map(|s| s.0).collect::<Vec<_>>().into(),
					style:     None,
					alignment: Default::default(),
				})
			})?,
		)
	}

	pub fn render(self, buf: &mut ratatui::buffer::Buffer) {
		let mut p = ratatui::widgets::Paragraph::new(self.text);
		if let Some(style) = self.style {
			p = p.style(style);
		}

		p = p.alignment(self.alignment);
		p.render(self.area, buf)
	}
}

impl<'lua> FromLua<'lua> for Paragraph {
	fn from_lua(value: Value<'lua>, _: &'lua Lua) -> mlua::Result<Self> {
		match value {
			Value::UserData(ud) => Ok(ud.borrow::<Self>()?.clone()),
			_ => Err(mlua::Error::FromLuaConversionError {
				from:    value.type_name(),
				to:      "Paragraph",
				message: Some("expected a Paragraph".to_string()),
			}),
		}
	}
}

impl UserData for Paragraph {
	fn add_methods<'lua, M: mlua::UserDataMethods<'lua, Self>>(methods: &mut M) {
		methods.add_function("style", |_, (ud, style): (AnyUserData, Style)| {
			ud.borrow_mut::<Self>()?.style = Some(style.0);
			Ok(ud)
		});
		methods.add_function("align", |_, (ud, align): (AnyUserData, u8)| {
			ud.borrow_mut::<Self>()?.alignment = match align {
				1 => ratatui::prelude::Alignment::Center,
				2 => ratatui::prelude::Alignment::Right,
				_ => ratatui::prelude::Alignment::Left,
			};
			Ok(ud)
		});
	}
}
